[RISCV] Lower llvm.clear_cache to __riscv_flush_icache for glibc targets

rofirrim opened the pull request:
We are working on Fortran applications on RISC-V Linux and some of them rely on passing a pointer to an internal (aka nested) function that uses the enclosing context (e.g., a variable of the enclosing function). [Flang uses trampolines](https://flang.llvm.org/docs/InternalProcedureTrampolines.html) which rely on llvm.init.trampoline and llvm.adjust.trampoline. Those rely on having an executable stack.

This change is a preliminary step to support trampolines on RISC-V.

In this change we lower `llvm.clear_cache` intrinsic on glibc targets to `__riscv_flush_icache` ( https://www.gnu.org/software/libc/manual/html_node/RISC_002dV.html ) which is what GCC is currently doing (https://www.godbolt.org/z/qsd1P3fYT).

I could not find where it is specified that RISC-V on glibc targets must call `__riscv_flush_icache` so before landing this we may want to address this issue.

arsenm commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -4764,8 +4764,15 @@ class TargetLowering : public TargetLoweringBase {
+  /// signature. When returning false, the lowering will invoke
+  /// getClearCacheBuiltinName.
+  virtual bool isClearCacheBuiltinTargetSpecific() const { return false; }
+
   /// Return the builtin name for the __builtin___clear_cache intrinsic
-  /// Default is to invoke the clear cache library call
+  /// This is only used if isClearCacheBuiltinTargetSpecific returns false.
+  /// If nullptr is returned, the builtin is lowered to no code.
   virtual const char * getClearCacheBuiltinName() const {
```
This was already a bad API and this is adding more parallel knobs. Instead, we should have an ISD node for the cache clear that legalizes in the normal way rather than trying to special case it in SelectionDAGBuilder. We can then delete these special case hooks

rofirrim replied:
Fair, let's discuss the groundwork at https://github.com/llvm/llvm-project/pull/93795

wangpc-pp commented:
Just FYI, we have implemented `__clear_cache` via calling `__NR_riscv_flush_icache` in `compiler-rt`:
https://github.com/llvm/llvm-project/blob/5c7c1f6aba467b1c5a51d58fc591314ecd7147ce/compiler-rt/lib/builtins/clear_cache.c#L184-L194
So we can link it by specifying ` --rtlib=compiler-rt`.
For `libgcc`, `__clear_cache` does nothing.
For glibc's `__riscv_flush_icache` implementation, it just calls the `__NR_riscv_flush_icache` (the same as `compiler-rt`'s `__clear_cache` implementation), and there may be vDSO acceleration.

rofirrim committed:
[RISCV] Custom lower llvm.clear_cache on glibc targets

Typically __clear_cache is used but on these targets __riscv_flush_icache
is used instead. Because it also has a different signature we need
custom lowering.

arsenm commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -7117,7 +7122,48 @@ SDValue RISCVTargetLowering::LowerOperation(SDValue Op,
+  EVT Ty = getPointerTy(DAG.getDataLayout());
+  CLI.setDebugLoc(DL).setChain(InChain).setLibCallee(
+      CallingConv::C, Type::getVoidTy(*DAG.getContext()),
+      DAG.getExternalSymbol("__riscv_flush_icache", Ty), std::move(Args));
```
Can't you just override the libfunc name with setLibcallName and use the default call lowering?

rofirrim commented:
> Just FYI, we have implemented `__clear_cache` via calling `__NR_riscv_flush_icache` in `compiler-rt`:

Thanks for the update. Do you use `compiler-rt` even on glibc targets? Not sure if there is a way to know we're on a glibc target **and** we are using `libgcc`  for the compiler runtime (which would indeed have to call `__riscv_flush_icache`).

arsenm commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -7117,7 +7122,48 @@ SDValue RISCVTargetLowering::LowerOperation(SDValue Op,
   case ISD::EXPERIMENTAL_VP_REVERSE:
     return lowerVPReverseExperimental(Op, DAG);
+  case ISD::CLEAR_CACHE: {
+    assert(Subtarget.getTargetTriple().isOSGlibc() &&
```
Should get from target machine, not subtarget

arsenm commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -7117,7 +7122,48 @@ SDValue RISCVTargetLowering::LowerOperation(SDValue Op,
+      CallingConv::C, Type::getVoidTy(*DAG.getContext()),
+      DAG.getExternalSymbol("__riscv_flush_icache", Ty), std::move(Args));
+
+  std::pair<SDValue, SDValue> CallResult = LowerCallTo(CLI);
```
Using TLI::makeLibCall may be a bit shorter than all this

rofirrim replied:
`makeLibCall` expects a `RTLIB::Libcall` and I could use `RTLIB::CLEAR_CACHE` after renaming it. However, I feel uneasy renaming `__clear_cache` into `__riscv_flush_icache` given that they have different signatures.

arsenm replied:
Alternatively could define this in RuntimeLibcalls

rofirrim replied:
Double checking here: you mean adding a new `HANDLE_LIBCALL` in `RuntimeLibcalls.def`, right? 

(I took a look and they seem largely target independent hence my question although there are some PowerPC specific bits as well.)

arsenm replied:
Yes. The entire concept of libcalls is target dependent, so this list is theoretically just a union of all target dependent libcalls

wangpc-pp commented:
> > Just FYI, we have implemented `__clear_cache` via calling `__NR_riscv_flush_icache` in `compiler-rt`:
> 
> Thanks for the update. Do you use `compiler-rt` even on glibc targets?

No I don't use `compiler-rt` on glibc. 

> Not sure if there is a way to know we're on a glibc target **and** we are using `libgcc` for the compiler runtime (which would indeed have to call `__riscv_flush_icache`).

We can know it in Clang via the `RuntimeLibType` we use, but I think this information won't be passed to LLVM's CodeGen part.

Maybe another way to fix this issue is implementing `__clear_cache` in libgcc. cc @kito-cheng

rofirrim committed:
Get the TargetTriple directly from the TargetMachine

rofirrim committed:
Add a runtime call for __riscv_flush_icache

This allows us to make the code much simpler.

rofirrim commented:
@kito-cheng @wangpc-pp I currently gated this to glibc.

However, investigating it further this seems a Linux ABI specific thing ( https://docs.kernel.org/arch/riscv/cmodx.html )  supported by both glibc and musl ( https://git.musl-libc.org/cgit/musl/tree/src/linux/cache.c#n39 ).

It looks to me it makes more sense to gate this to just Linux and not glibc (which would also include things like Hurd).

Do you agree?

wangpc-pp commented:
> @kito-cheng @wangpc-pp I currently gated this to glibc.
> 
> However, investigating it further this seems a Linux ABI specific thing ( https://docs.kernel.org/arch/riscv/cmodx.html ) supported by both glibc and musl ( https://git.musl-libc.org/cgit/musl/tree/src/linux/cache.c#n39 ).
> 
> It looks to me it makes more sense to gate this to just Linux and not glibc (which would also include things like Hurd).
> 
> Do you agree?

Yeah, make sense to me. Maybe we should write it down in psABI or elsewhere? @kito-cheng

kito-cheng commented:
I agree that should just gate with linux rather than glibc, but I am incline do not including OS specific stuff within psABI if possible (although we already has document `wchat_t` in psABI)

rofirrim committed:
Gate lowering to Linux instead of Glibc

kito-cheng reviewed (approved):
LGTM :)

dyung commented:
Our Windows build also seems to be seeing similar failures as the `llvm-nvptx-nvidia-win` builder https://lab.llvm.org/buildbot/#/builders/46/builds/312

rofirrim commented:
> Our Windows build also seems to be seeing similar failures as the `llvm-nvptx-nvidia-win` builder https://lab.llvm.org/buildbot/#/builders/46/builds/312

I'll take a look. Thanks for the heads up.

rofirrim commented:
@dyung if you can revert the change that'd be helpful, so I can investigate. Thanks!

dyung commented:
> @dyung if you can revert the change that'd be helpful, so I can investigate. Thanks!

Oddly it no longer seems to be failing on the bot and I don’t think there was a revert. Perhaps it was some unrelated change? I’ll try to verify today whether it was your change or not that caused the test failures.

rofirrim commented:
> > @dyung if you can revert the change that'd be helpful, so I can investigate. Thanks!
> 
> Oddly it no longer seems to be failing on the bot and I don’t think there was a revert. Perhaps it was some unrelated change? I’ll try to verify today whether it was your change or not that caused the test failures.

Oh, good to know. Thanks a lot @dyung !
